core/matrix/dense.cpp

```diff
@@ -1080,48 +1094,174 @@ void Dense<ValueType>::conj_transpose(ptr_param<Dense<ValueType>> output) const
+        break;
+    case permute_mode::inverse_rows:
+        exec->run(dense::make_inverse_row_permute(
+            local_perm->get_const_permutation(), this, local_output.get()));
+        break;
+    case permute_mode::inverse_columns:
+        exec->run(dense::make_inverse_col_permute(
+            local_perm->get_const_permutation(), this, local_output.get()));
+        break;
+    case permute_mode::inverse_symmetric:
+        exec->run(dense::make_inv_symm_permute(
+            local_perm->get_const_permutation(), this, local_output.get()));
+        break;
+    default:
+        GKO_ASSERT(false);  // cannot happen
+    }
 }
 
 
 template <typename ValueType>
 template <typename IndexType>
-void Dense<ValueType>::inverse_permute_impl(
-    const array<IndexType>* permutation_indices, Dense<ValueType>* output) const
+void Dense<ValueType>::permute_impl(
+    const Permutation<IndexType>* row_permutation,
+    const Permutation<IndexType>* col_permutation, bool invert,
+    Dense* output) const
 {
-    GKO_ASSERT_IS_SQUARE_MATRIX(this);
-    GKO_ASSERT_EQUAL_DIMENSIONS(this, output);
-    GKO_ASSERT_EQ(permutation_indices->get_num_elems(), this->get_size()[0]);
     auto exec = this->get_executor();
-
-    exec->run(dense::make_inv_symm_permute(
-        make_temporary_clone(exec, permutation_indices).get(), this,
-        make_temporary_output_clone(exec, output).get()));
+    auto size = this->get_size();
+    GKO_ASSERT_EQUAL_DIMENSIONS(this, output);
+    GKO_ASSERT_EQ(size[0], row_permutation->get_size()[0]);
+    GKO_ASSERT_EQ(size[1], col_permutation->get_size()[0]);
+    auto local_output = make_temporary_output_clone(exec, output);
+    auto local_row_perm = make_temporary_clone(exec, row_permutation);
+    auto local_col_perm = make_temporary_clone(exec, col_permutation);
+    if (invert) {
+        exec->run(dense::make_inv_nonsymm_permute(
+            local_row_perm->get_const_permutation(),
+            local_col_perm->get_const_permutation(), this, local_output.get()));
+    } else {
+        exec->run(dense::make_nonsymm_permute(
+            local_row_perm->get_const_permutation(),
+            local_col_perm->get_const_permutation(), this, local_output.get()));
+    }
 }
 
 
 template <typename ValueType>
 template <typename IndexType>
-void Dense<ValueType>::row_permute_impl(
-    const array<IndexType>* permutation_indices, Dense<ValueType>* output) const
+void Dense<ValueType>::scale_permute_impl(
+    const ScaledPermutation<ValueType, IndexType>* permutation,
+    permute_mode mode, Dense* output) const
 {
-    GKO_ASSERT_EQ(permutation_indices->get_num_elems(), this->get_size()[0]);
+    const auto exec = this->get_executor();
+    const auto size = this->get_size();
     GKO_ASSERT_EQUAL_DIMENSIONS(this, output);
-    auto exec = this->get_executor();
+    if ((mode & permute_mode::symmetric) == permute_mode::none) {
+        output->copy_from(this);
+        return;
+    }
```

Comment:
I assume symmetric is the full bit set of all possible settings but it is not obvious here I think.
Could we just check permute_mode::none without the mask? the user input more than the bitset will also throw the error in the end.

Reply:
no, symmetric is rows | columns, so this checks that we are permuting both at the same time